cfme/tests/infrastructure/test_snapshot.py

```diff
@@ -173,6 +173,51 @@ def test_verify_revert_snapshot(full_test_vm, provider, soft_assert, register_ev
     ssh_client.close()
 
 
+def setup_snapshot_env(test_vm, memory):
+    logger.info("Starting snapshot setup")
+    snapshot1 = new_snapshot(test_vm, memory=memory)
+    snapshot1.create()
+    snapshot2 = new_snapshot(test_vm, memory=memory)
+    snapshot2.create()
+    snapshot1.revert_to()
+    wait_for(lambda: snapshot1.active is True,
+             num_sec=300, delay=20, fail_func=sel.refresh)
+
+
+@pytest.mark.parametrize("parent_vm", ["on_with_memory", "on_without_memory", "off"])
+@pytest.mark.uncollectif(lambda provider: not provider.one_of(VMwareProvider))
+def test_verify_vm_state_revert_snapshot(provider, parent_vm, request, small_test_vm):
+    """
+    test vm state after revert snapshot with parent vm:
+     - powered on and includes memory
+     - powered on without memory
+     - powered off
+
+    vm state after revert should be:
+     - powered on
+     - powered off
+     - powered off
+    """
+    if parent_vm == "on_with_memory":
+        logger.info("Making sure VM is ON")
+        small_test_vm.power_control_from_cfme(option=small_test_vm.POWER_ON, cancel=False)
+        provider.mgmt.wait_vm_steady(small_test_vm.name)
+        setup_snapshot_env(small_test_vm, True)
+        assert small_test_vm.provider.mgmt.is_vm_running(small_test_vm.name)
+    elif parent_vm == "on_without_memory":
+        logger.info("Making sure VM is ON")
+        small_test_vm.power_control_from_cfme(option=small_test_vm.POWER_ON, cancel=False)
+        provider.mgmt.wait_vm_steady(small_test_vm.name)
+        setup_snapshot_env(small_test_vm, False)
+        assert not small_test_vm.provider.mgmt.is_vm_running(small_test_vm.name)
+    else:
+        logger.info("Making sure VM is OFF")
+        small_test_vm.power_control_from_cfme(option=small_test_vm.POWER_OFF, cancel=False)
+        provider.mgmt.wait_vm_steady(small_test_vm.name)
+        setup_snapshot_env(small_test_vm, False)
+        assert not small_test_vm.provider.mgmt.is_vm_running(small_test_vm.name)
```

Comment:
This could be generalized like this instead of the repeating code in the if branches:
```python
power = small_test_vm.POWER_ON if parent_vm.startswith('on') else small_test_vm.POWER_OFF
snap_env = 'with_memory' in parent_vm

small_test_vm.power_control_from_cfme(option=power, cancel=False)
provider.mgmt.wait_vm_steady(small_test_vm.name)
setup_snapshot_env(small_test_vm, snap_env)
assert bool(small_test_vm.provider.mgmt.is_vm_running(small_test_vm.name)) == snap_env
```

What do you think?

Reply:
This indeed looks much better.